zallet/src/components/json_rpc/balance.rs

```
@@ -0,0 +1,293 @@
+        )
+        // NB: negative valueBalanceSapling "takes" money from the transparent value pool just as outputs do
+        .chain((-tx.sapling_value_balance()).try_into().ok())
+        // NB: negative valueBalanceOrchard "takes" money from the transparent value pool just as outputs do
+        .chain(
+            tx.orchard_bundle()
+                .and_then(|b| (-*b.value_balance()).try_into().ok()),
+        )
+        .chain(tx.sprout_bundle().into_iter().flat_map(|b| {
+            b.joinsplits
+                .iter()
+                // Consensus rule: either `vpub_old` or `vpub_new` MUST be zero.
+                // Therefore if `JsDescription::net_value() <= 0`, it is equal to
+                // `-vpub_old`.
+                .flat_map(|jsdesc| (-jsdesc.net_value()).try_into().ok())
+        }))
+        .sum()
+}
+
+/// Equivalent to [`CWalletTx::GetDebit`] in `zcashd`.
+///
+/// [`CWalletTx::GetDebit`]: https://github.com/zcash/zcash/blob/2352fbc1ed650ac4369006bea11f7f20ee046b84/src/wallet/wallet.cpp#L4822
+pub(super) fn wtx_get_debit(
+    wallet: &DbConnection,
+    tx: &Transaction,
+    is_mine: impl Fn(&DbConnection, &TxOut) -> Result<bool, SqliteClientError>,
+) -> Result<Option<Zatoshis>, SqliteClientError> {
+    match tx.transparent_bundle() {
+        None => Ok(Some(Zatoshis::ZERO)),
+        Some(bundle) if bundle.vin.is_empty() => Ok(Some(Zatoshis::ZERO)),
+        // Equivalent to `CWallet::GetDebit(CTransaction)` in `zcashd`.
+        Some(bundle) => {
+            let mut acc = Some(Zatoshis::ZERO);
+            for txin in &bundle.vin {
+                // Equivalent to `CWallet::GetDebit(CTxIn)` in `zcashd`.
+                if let Some(txout) = wallet
+                    .get_transaction(*txin.prevout.txid())?
+                    .as_ref()
+                    .and_then(|prev_tx| prev_tx.transparent_bundle())
+                    .and_then(|bundle| bundle.vout.get(txin.prevout.n() as usize))
```

bug: This seems unusable in the context of `zallet`; in zcashd, this behavior was okay because you could be confident that you had access to the transaction inputs; here, this will just silently ignore any value for inputs for which you don't have the full transaction.

---

Can you provide a suggestion on what to do here?